beacon-chain/p2p/peers/score_block_providers.go

```diff
@@ -0,0 +1,178 @@
+)
+
+// BlockProviderScorer represents block provider scoring service.
+type BlockProviderScorer struct {
+	ctx    context.Context
+	config *BlockProviderScorerConfig
+	store  *peerDataStore
+	// maxProcessedBlocks defines maximum number of processed blocks attained by some peer during
+	// the lifetime of a service. This is used to gauge relative performance of other peers.
+	maxProcessedBlocks uint64
+}
+
+// BlockProviderScorerConfig holds configuration parameters for block providers scoring service.
+type BlockProviderScorerConfig struct {
+	// ProcessedBatchWeight defines a reward for a single processed batch of blocks.
+	ProcessedBatchWeight float64
+	// DecayInterval defines how often stats should be decayed.
+	DecayInterval time.Duration
+	// Decay specifies number of blocks subtracted from stats on each decay step.
+	Decay uint64
+}
+
+// newBlockProviderScorer creates block provider scoring service.
+func newBlockProviderScorer(
+	ctx context.Context, store *peerDataStore, config *BlockProviderScorerConfig) *BlockProviderScorer {
+	if config == nil {
+		config = &BlockProviderScorerConfig{}
+	}
+	scorer := &BlockProviderScorer{
+		ctx:                ctx,
+		config:             config,
+		store:              store,
+		maxProcessedBlocks: uint64(flags.Get().BlockBatchLimit),
+	}
+	if scorer.config.ProcessedBatchWeight == 0.0 {
+		scorer.config.ProcessedBatchWeight = DefaultBlockProviderProcessedBatchWeight
+	}
+	if scorer.config.DecayInterval == 0 {
+		scorer.config.DecayInterval = DefaultBlockProviderDecayInterval
+	}
+	if scorer.config.Decay == 0 {
+		scorer.config.Decay = uint64(flags.Get().BlockBatchLimit)
+	}
+	return scorer
+}
+
+// Score calculates and returns block provider score.
+func (s *BlockProviderScorer) Score(pid peer.ID) float64 {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.score(pid)
+}
+
+// score is a lock-free version of Score.
+func (s *BlockProviderScorer) score(pid peer.ID) float64 {
+	score := float64(0)
+	peerData, ok := s.store.peers[pid]
+	if !ok {
+		return score
+	}
+	batchSize := uint64(flags.Get().BlockBatchLimit)
+	if batchSize > 0 {
+		processedBatches := float64(peerData.processedBlocks / batchSize)
+		score += processedBatches * s.config.ProcessedBatchWeight
+	}
+	return math.Round(score*ScoreRoundingFactor) / ScoreRoundingFactor
+}
+
+// Params exposes scorer's parameters.
+func (s *BlockProviderScorer) Params() *BlockProviderScorerConfig {
+	return s.config
+}
+
+// IncrementProcessedBlocks increments the number of blocks that have been successfully processed.
+func (s *BlockProviderScorer) IncrementProcessedBlocks(pid peer.ID, cnt uint64) {
+	s.store.Lock()
+	defer s.store.Unlock()
+
+	if cnt <= 0 {
+		return
+	}
+	if _, ok := s.store.peers[pid]; !ok {
+		s.store.peers[pid] = &peerData{}
+	}
+	s.store.peers[pid].processedBlocks += cnt
+	if s.store.peers[pid].processedBlocks > s.maxProcessedBlocks {
+		s.maxProcessedBlocks = s.store.peers[pid].processedBlocks
```

Comment:
~is this correct ? shouldn't this be the other way around and max processed blocks is static while we limit our peer's `processedBlocks` value.~

Nvm, mistook the purpose of `maxProcessedBlocks`, can you change the name to something better ? It is confusing as from the name I assume the value is a static limit, rather than a lifetime gauge for blocks processed.

Reply:
Done!